File: code/1018-Binary Prefix Divisible By 5/1018-Binary Prefix Divisible By 5.py
# # Given an array A of 0s and 1s, consider N_i: the i-th subarray from A[0] to A[i] interpreted as a binary number (from most-significant-bit to least-significant-bit.)

# Return a list of booleans answer, where answer[i] is true if and only if N_i is divisible by 5.

# Example 1:

# Input: [0,1,1]
# Output: [true,false,false]
# Explanation:
# The input numbers in binary are 0, 01, 011; which are 0, 1, and 3 in base-10.  Only the first number is divisible by 5, so answer[0] is true.
# Example 2:

# Input: [1,1,1]
# Output: [false,false,false]
# Example 3:

# Input: [0,1,1,1,1,1]
# Output: [true,false,false,false,true,false]
# Example 4:

# Input: [1,1,1,0,1]
# Output: [false,false,false,false,false]
#  

# Note:

# 1 <= A.length <= 30000
# A[i] is 0 or 1



class Solution(object):
    def prefixesDivBy5(self, A):
        """
        :type A: List[int]
        :rtype: List[bool]
        """
        res = []
        digit = 0
        for i, item in enumerate(A):
            digit = digit * 2 + item
            res.append(digit % 5 == 0)
            digit = digit % 5
        return res

if __name__ == '__main__':
    s = Solution()
    print(s.prefixesDivBy5([1,1,1]))
    print(s.prefixesDivBy5([0,1,1,1,1,1]))
    print(s.prefixesDivBy5([1,1,1,0,1]))

File: code/1018-Binary Prefix Divisible By 5/1018-Binary Prefix Divisible By 5.cpp
/*
LeetCode 1018 - Binary Prefix Divisible By 5

解题步骤（自动整理）
1. 先把题目目标拆成可操作的子任务（比较/统计/构造等）
2. 选择合适的数据结构并按一次遍历或分治步骤实现核心逻辑
3. 补齐边界条件（空输入、单元素、重复元素等）并返回结果
*/
/*
Given an array A of 0s and 1s, consider N_i: the i-th subarray from A[0] to A[i] interpreted as a binary number (from most-significant-bit to least-significant-bit.)

Return a list of booleans answer, where answer[i] is true if and only if N_i is divisible by 5.

Example 1:

Input: [0,1,1]
Output: [true,false,false]
Explanation:
The input numbers in binary are 0, 01, 011; which are 0, 1, and 3 in base-10.  Only the first number is divisible by 5, so answer[0] is true.
Example 2:

Input: [1,1,1]
Output: [false,false,false]
Example 3:

Input: [0,1,1,1,1,1]
Output: [true,false,false,false,true,false]
Example 4:

Input: [1,1,1,0,1]
Output: [false,false,false,false,false]


Note:

1 <= A.length <= 30000
A[i] is 0 or 1
*/

#include "vector"

using namespace std;

class Solution {
public:
    vector<bool> prefixesDivBy5(vector<int>& A) {
        vector<bool> res;
        int digit = 0, n = A.size();
        for (int i = 0; i < n; i ++) {
            digit = digit * 2 + A[i];
            res.push_back(digit % 5 == 0);
            digit = digit % 5;
        }
        return res;

    }
};
